### PR TITLE
[None][fix] Restore grammar state after SA, DFlash and PARD verification

### PR DESCRIPTION
**Pull request by brnguyen2**
## Dev Engineer Review

- Adds guided-decoder acceptance tracking and rejected-token rollback.
- Adds post-verification rollback for `SAWorker`, `DFlashWorker`, and `PARDWorker`.
- Clamps accepted-token counts and preserves CUDA graph synchronization.
- Key risks include matcher restoration, rejected EOS tokens, terminated matchers, and zero-draft paths.
- No review severity findings were supplied.

## QA Engineer Review

- Adds three CUDA-gated test modules with 8 test functions and 16 parametrized cases.
- Covers matcher restoration, terminal drafts, CUDA stream and graph behavior, batch-position indexing, accepted-count forwarding, worker call ordering, and zero-draft paths.
- The tests are included by the broad `unittest/_torch/speculative/hw_agnostic` entries in `l0_cpu.yml` and `l0_h100.yml`.
- The author reports 16 passed tests with no skips or failures. Reported CI pipelines failed, so broader CI completion is not an all-clear.
- Coverage verdict: **sufficient**.

## Per-File QA Perspective

- `tensorrt_llm/_torch/pyexecutor/guided_decoder.py`: Verify the new acceptance APIs, matcher restoration, rejected-token handling, count clamping, and CUDA event synchronization.
- `tensorrt_llm/_torch/speculative/interface.py`: Verify rollback guards and post-verification call ordering.
- `tensorrt_llm/_torch/speculative/dflash.py`: Verify rollback after KV and attention-state restoration.
- `tensorrt_llm/_torch/speculative/pard.py`: Verify rollback uses per-row accepted-token counts.
- `tensorrt_llm/_torch/speculative/sa_worker.py`: Verify rollback uses verification counts and is skipped when drafting is disabled.
- `tests/unittest/_torch/speculative/hw_agnostic/test_dflash_guided_rollback.py`: Covers DFlash rollback and zero-draft behavior. Included through the broad `hw_agnostic` CI test-list entries.
- `tests/unittest/_torch/speculative/hw_agnostic/test_pard_guided_rollback.py`: Covers PARD rollback, per-row counts, and zero-draft behavior. Included through the broad `hw_agnostic` CI test-list entries.
- `tests/unittest/_torch/speculative/hw_agnostic/test_sa_guided_rollback.py`: Covers matcher restoration, terminal drafts, CUDA graph replay, batch-position indexing, regression behavior, call ordering, and zero-draft behavior. Included through the broad `hw_agnostic` CI test-list entries.

## Motivation

`CapturableGuidedDecoder.execute` advances each request's grammar matcher through
the golden token and as many draft tokens as the grammar accepts, so that the
target step can mask every verified position at once. The rejected suffix has to
be rolled back before the next target step, or the matcher is left at a state
that includes tokens the target never accepted.

Drafters that run a draft forward pass do this inside
`execute_draft_batch(draft_step=0)`, which calls `rollback_rejected_tokens`.
Three workers never enter that loop:

- `SAWorker` drafts from the suffix automaton,
- `DFlashWorker` and `PARDWorker` draft with their own kernels,

so after verification their matchers still contain the rejected draft tokens.
The next target step then either masks from the wrong grammar state, or fails to
mask at all when the golden token is not acceptable from that state — in which
case constrained decoding can emit tokens the grammar forbids, with no error.

A rejected EOS makes this worse: the matcher can be left terminated, after which
no bitmask is produced at all.

## Summary

- Add `CapturableGuidedDecoder.rollback_rejected_batch(num_accepted_tokens)`,
  plus the `add_accepted_batch` / `fetch_accepted_batch` pair it needs. These
  mirror the existing `add_draft_batch` / `fetch_draft_batch` /
  `execute_draft_batch(draft_step=0)` path, minus the parts that only the draft
  loop needs (new-token column, `cast_to_draft`, bitmask build).
- Add `SpecWorkerBase._rollback_guided_decoder_after_verify`, a sibling of the
  existing `_execute_guided_decoder_if_present`, and call it from `SAWorker`,
  `DFlashWorker` and `PARDWorker` after drafting. The requirement is a property
  of the worker base class -- "a drafter with no per-step drafted-logits loop
  must roll matchers back after verification" -- so the next one-shot drafter
  inherits it rather than having to rediscover it.
  Placement is deliberate: the rollback enqueues CUDA host callbacks that take
  the GIL, so it must come after native drafting calls that may synchronize the
  stream while holding it.
- It deliberately does NOT live in `_prepare_next_new_tokens`, the one call all
  three workers already share at exactly that point. `MTPWorker`,
  `Eagle3OneModelWorker`, `DSv4DSparkWorker`, `DraftTargetOneModelWorker` and
  `MTPDynamicTreeWorker` call that too, and they already roll back via
  `execute_draft_batch(draft_step=0)` -- they would roll back twice.
- The verification count is clamped to `num_advanced_tokens[slot]` before it
  becomes `num_accepted_draft_tokens`, the same clamp `fetch_draft_batch`
  already applies. The target can accept tokens past the point where the matcher
  stopped advancing (EOS inside the drafts, a terminated matcher, or an
  unconstrained tail); without the clamp `_rollback_rejected_tokens` would
  compute a negative rollback length and raise.
- The side stream is rejoined via `bitmask_event` so the sequence is capturable
  in a CUDA graph.

State after the call: the matcher sits exactly at the last accepted token, which
is what the next target step's `accept_token(new_token)` expects. Requests with
no guided-decoding params, no draft tokens, or `num_advanced_tokens == 0` are
skipped by the existing guards, and a zero runtime draft length returns through
`skip_drafting` before any matcher advance, so there is nothing to roll back.

## Test Coverage

Three new modules under `tests/unittest/_torch/speculative/hw_agnostic/`:
`test_sa_guided_rollback.py`, `test_dflash_guided_rollback.py`,
`test_pard_guided_rollback.py` — 16 cases total, covering:

- every acceptance length from 1 to `max_num_draft_tokens + 1`,
- a rejected terminal (EOS) draft token, including re-accepting EOS after
  rollback,
- acceptance counts beyond the point where the matcher stopped advancing,
- CUDA graph capture and replay of the rollback,
- `execute` → draft → `rollback_rejected_batch` call ordering in each of the
  three workers,
- the zero-runtime-draft-length path, which must not roll back,
- a control case showing the pre-fix behaviour: without the rollback the next
  verification cannot accept the golden token and applies no mask at all.

They use a 16-symbol synthetic vocabulary and the built-in JSON grammar. No
model weights, no checkpoint, no network. The
`unittest/_torch/speculative/hw_agnostic` directory is already listed in
`l0_cpu.yml` and `l0_h100.yml`, so no test-list change is needed; the cases are
`skipif`-guarded on `torch.cuda.is_available()` and skip on the CPU job, as 11
of the 30 pre-existing modules in that directory already do.

## Validation status

The 16 cases were run on an SM100 build of this exact source revision:
**16 passed, 0 failed, 0 skipped**, pytest rc 0, with collection verified to
list exactly the 16 node IDs first. Zero skips matters here: every case is
`skipif`-guarded on `torch.cuda.is_available()`, so a skipped run would have
proven nothing.

That run includes the three per-worker call-ordering cases, which exercise the
new `SpecWorkerBase` hook end to end rather than the old per-worker call sites.

`pre-commit run` is clean on all changed files and the commit is signed off.

## PR Checklist

- [x] Commit signed off (DCO)
- [x] `pre-commit` clean on all changed files
- [x] Unit tests run on a GPU against a matching build (16 passed / 0 skipped)
- [x] `/bot run`

### CODE REVIEW

File: tests/unittest/_torch/speculative/hw_agnostic/test_dflash_guided_rollback.py [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+def test_dflash_worker_rolls_back_after_verification(monkeypatch):
+    worker = DFlashWorker(DFlashDecodingConfig(max_draft_len=K), Mapping())
+    guided = _RecordingGuidedDecoder()
+    worker.guided_decoder = guided
+
+    # A context-only batch skips the draft forward but still runs the guided
+    # mask and the verification, so the call order is pinned without loading a
+    # draft model. The buffer bookkeeping around them is stubbed out.
+    batch_size = num_contexts = 2
+    meta, attn, *inputs = _worker_inputs(batch_size, num_contexts, runtime_draft_len=K)
```
[the remark this answers is not in the document]
Comment (brnguyen2) [reply tests/unittest/_torch/speculative/hw_agnostic/test_dflash_guided_rollback.py:92]: Addressed: both tests now mock a distinct per-row accepted-count (`[K+1, 1]` — one row accepts the whole draft, the other only the golden token) instead of a uniform all-ones tensor, so the equality against the tensor reaching `rollback_rejected_batch` now pins per-row forwarding rather than "ones equal ones".

I kept the batch context-only (`num_gens == 0`) deliberately. A real generation row needs a loaded draft model — the `num_gens > 0` branch calls `draft_model.dflash_forward` (and the PARD equivalent) — which these call-ordering tests avoid by design. It also would not add coverage at the seam under test: the worker passes `num_accepted_tokens` whole to `_rollback_guided_decoder_after_verify` (dflash.py:1191 / pard.py:346), which passes it whole to `rollback_rejected_batch` (interface.py:2697); nothing between the mock and the rollback slices by `num_contexts`, so a context row and a generation row are indistinguishable there. Generation-count derivation inside `sample_and_accept_draft_tokens` is a separate concern, out of scope for these forwarding tests.

File: tensorrt_llm/_torch/pyexecutor/guided_decoder.py [hunk 2]
```diff
@@ -628,6 +631,58 @@ def fetch_draft_batch(self, draft_step: int = 0) -> None:
...
+        num_accepted_tokens_list = self.num_accepted_tokens[:batch_size].tolist(
+        )
+        for i, req in enumerate(self.requests_hostfunc.requests):
+            if req.guided_decoding_params is None or (slot :=
+                                                      req.seq_slot) is None:
+                continue
+            # Verification can accept tokens beyond a terminal draft token,
+            # where the matcher stopped advancing (see fetch_draft_batch).
+            req.num_accepted_draft_tokens = min(
+                num_accepted_tokens_list[i], self.num_advanced_tokens[slot]) - 1
```
[the remark this answers is not in the document]
Comment (brnguyen2) [reply tensorrt_llm/_torch/pyexecutor/guided_decoder.py:659]: Agreed that batch size 1 hides this: with one request the enumerate index and `seq_slot` always coincide. `fetch_accepted_batch` does index the count list by batch position `i` and uses `seq_slot` only for `num_advanced_tokens` and matcher state, which is correct but was untested. Added `test_rollback_indexes_counts_by_batch_position_not_seq_slot` in d217a3a241: two guided rows on slots 0 and 1 in reversed batch order, an unguided row, and distinct per-position counts, asserting each guided request's `num_accepted_draft_tokens` and that matcher state followed the slot.

**Review by zhaoyuanh-nvidia — APPROVED** [review]
